Agents/AverageSquareMetrePriceAgent/avgsqmpriceagent/agent/avgprice_estimation.py

```diff
@@ -0,0 +1,285 @@
+################################################
+# Authors: Markus Hofmeister (mh807@cam.ac.uk) #    
+# Date: 21 Oct 2022                            #
+################################################
+
+# The purpose of this module is to instantiate/update the average square metre 
+# price of properties for a postcode based on instantiated HM Land Registry's Price
+# Paid Data transactions in the KG (using asynchronous derivation framework)
+
+import uuid
+import pandas as pd
+
+from pyderivationagent import DerivationAgent
+from pyderivationagent import DerivationInputs
+from pyderivationagent import DerivationOutputs
+from rdflib import Graph
+
+from avgsqmpriceagent.datamodel.iris import *
+from avgsqmpriceagent.datamodel.data import TIME_FORMAT_LONG, TIME_FORMAT_SHORT
+from avgsqmpriceagent.errorhandling.exceptions import TSException
+from avgsqmpriceagent.kg_operations.kgclient import KGClient
+from avgsqmpriceagent.kg_operations.tsclient import TSClient
+
+
+class AvgSqmPriceAgent(DerivationAgent):
+
+    def __init__(self, **kwargs):
+        # Set transaction threshold for average calculation
+        self.threshold = kwargs.pop('threshold')
+        # Initialise DerivationAgent parent instance
+        super().__init__(**kwargs)
+
+        # Initialise the Sparql_client (with defaults specified in environment variables)
+        self.sparql_client = self.get_sparql_client(KGClient)
+        
+
+    def agent_input_concepts(self) -> list:
+        # Please note: Declared inputs/outputs need proper instantiation incl. 
+        #              RDF TYPE declarations in the KG for the derivation to work
+        return [OBE_POSTALCODE, OBE_PROPERTY_PRICE_INDEX,
+                LRPPI_TRANSACTION_RECORD]
+
+
+    def agent_output_concepts(self) -> list:
+        # Output concept (i.e. result) of the Derivation
+        return [OBE_AVERAGE_SM_PRICE]
+
+
+    def validate_inputs(self, http_request) -> bool:
+        # Validate completeness of received HTTP request (i.e. non-empty HTTP request, 
+        # contains derivationIRI, etc.) -> only relevant for synchronous derivation
+        return super().validate_inputs(http_request)
+
+
+    def validate_input_values(self, inputs):
+        """
+        Check whether received input values are suitable to perform average price
+        estimation. Throw exception if data is not suitable.
+        -> relevant for asynchronous derivation
+
+        Arguments:
+            inputs {dict} -- Dictionary of inputs with input concepts as keys and values as list
+
+        Returns:
+            postcode_iri {str}, ppi_iri {str}, tx_records {list}
+        """
+
+        #TODO: Potentially add derivationIRI to logger and exception message
```

Comment:
This is currently not easily possible, derivationIRI is hidden to `process_request_parameters`. Maybe we can change to `process_request_parameters(self, derivation_iri, derivation_inputs, derivation_outputs)` in the next version of pyderivationagent.

Reply:
Elaborated the comment in commit `[f813e37](https://github.com/cambridge-cares/TheWorldAvatar/pull/398/commits/f813e37ede1bb95d38dd8c67ef1b03cf675734fc)` to keep track of this

Reply:
With the latest version of [`pyderivationagent==1.3.0`](https://pypi.org/project/pyderivationagent/1.3.0/) (you also need `py4jps>=1.0.27`), you can now get the derivation IRI by `derivation_inputs.getDerivationIRI()`

Reply:
Included in [79b346f](https://github.com/cambridge-cares/TheWorldAvatar/pull/398/commits/79b346fa5a11c56f66186d806360318ce641955c)